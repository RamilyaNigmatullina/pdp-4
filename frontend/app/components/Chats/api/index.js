import axios from 'axios';

const MESSAGE_FIELDS = ['id', 'text', 'sender_id', 'created_at'];

export const createMessage = (chatId, message) => axios
  .post(`/api/v1/chats/${chatId}/messages`, { ...message })
  .then((response) => response.data);

export const fetchMessages = (chatId, createdAt) => axios
  .get(`/api/v1/chats/${chatId}/messages`, { params: { created_at_until: createdAt, fields: MESSAGE_FIELDS } })
  .then((response) => response.data);

export const readMessages = (chatId) => axios
  .post(`/api/v1/chats/${chatId}/reads`)
  .then((response) => response.data);

export const fetchUsersWithoutChat = () => axios
  .get('/api/v1/users')
  .then((response) => response.data);

export const createChat = (userId) => axios
  .post('/api/v1/chats', { chat: { user_id: userId } })
  .then((response) => response.data);
