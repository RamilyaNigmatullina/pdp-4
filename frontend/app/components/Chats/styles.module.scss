@import "~bootstrap/scss/functions";
@import "~bootstrap/scss/variables";
@import "~bootstrap/scss/mixins/breakpoints";

.chat {
  display: flex;
}

.chatsList {
  border-right: 1px solid #c4c4c4;

  @include media-breakpoint-down(md) {
    display: none;
  }
}

.chatItems {
  flex-direction: column;
  height: 550px;
  overflow-y: scroll;
  width: 350px;
}

.dialog {
  display: flex;
  flex-direction: column;
  height: calc(101vh - 64px);
  padding: 10px 0 1px 0;
  width: 100%;

  @include media-breakpoint-down(md) {
    height: calc(100vh - 56px);
  }
}
