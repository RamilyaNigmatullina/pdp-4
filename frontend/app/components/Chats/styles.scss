@import "~bootstrap/scss/functions";
@import "~bootstrap/scss/variables";
@import "~bootstrap/scss/mixins/breakpoints";

.chat {
  display: flex;
}

.dialog-container {
  display: flex;
  flex-direction: column;
  height: calc(103vh - 64px);
  padding: 10px 0;
  width: 100%;

  @include media-breakpoint-down(md) {
    height: calc(100vh - 56px);
  }
}

.chats-container {
  border-right: 1px solid #c4c4c4;
}
