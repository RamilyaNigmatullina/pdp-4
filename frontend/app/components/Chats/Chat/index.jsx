import React from 'react';
import PropTypes from 'prop-types';
import consumer from 'helpers/consumer';
import InfiniteScrolling from 'components/InfiniteScrolling';
import Message from './Message';
import MessageForm from './MessageForm';
import Interlocutor from './Interlocutor';
import { fetchMessages } from '../api/index';

class Chat extends React.Component {
  state = {
    isLastPage: false,
    messages: [],
    page: 1,
  };

  componentDidMount() {
    this.setState({
      subscription: this.subscribeToChannel(),
    });
  }

  componentDidUpdate(prevProps) {
    if (prevProps.chat.id !== this.props.chat.id) {
      this.setState({
        isLastPage: false,
        messages: [],
        page: 1,
        subscription: this.subscribeToChannel(),
      });
    }
  }

  subscribeToChannel = () => {
    const { subscription } = this.state;

    if (subscription) subscription.unsubscribe();

    return consumer.subscriptions.create({
      channel: 'ChatChannel',
      chat_id: this.props.chat.id,
    }, {
      received: this.handleReceived,
    });
  }

  renderMessage = (message) => {
    const { currentUser, chat } = this.props;
    const isCurrentUserMessage = this.isCurrentUserMessage(message);
    const avatarUrl = isCurrentUserMessage ? currentUser.avatar : chat.interlocutor.avatar;

    return (
      <Message
        avatarUrl={avatarUrl}
        isCurrentUserMessage={isCurrentUserMessage}
        key={message.id}
        message={message}
    />);
  }

  render() {
    const { messages, isLastPage } = this.state;
    const { chat } = this.props;

    return (
      <>
        { <Interlocutor interlocutor={chat.interlocutor} /> }

        <div className="messages">
          { <InfiniteScrolling
              isLastPage={isLastPage}
              items={messages}
              loadItems={this.handleLoadMessages}
              renderItem={this.renderMessage}
            /> }
        </div>

        { <MessageForm chatId={chat.id} /> }
      </>
    );
  }

  handleReceived = (message) => {
    if (this.isCurrentUserMessage(message)) this.handleMessageReceived(message);

    this.setState(({ messages }) => ({ messages: [message, ...messages] }), () => {
      const messageElement = document.getElementById(`message-${message.id}`);
      messageElement.scrollIntoView({ behavior: 'smooth' });
    });
  }

  handleMessageReceived = (message) => {
    const chat = { ...this.props.chat, last_message: message, unread_messages_count: 0 };
    const { onMessageReceived } = this.props;

    onMessageReceived(chat);
  }

  handleLoadMessages = () => {
    fetchMessages(this.props.chat.id, this.state.page)
      .then((data) => {
        this.setState(({ messages, page }) => ({
          isLastPage: !data.length,
          messages: [...messages, ...data],
          page: page + 1,
        }));
      });
  }

  isCurrentUserMessage = (message) => this.props.currentUser.id === message.sender_id
}

Chat.propTypes = {
  chat: PropTypes.object.isRequired,
  currentUser: PropTypes.object.isRequired,
  onMessageReceived: PropTypes.func.isRequired,
};

export default Chat;
