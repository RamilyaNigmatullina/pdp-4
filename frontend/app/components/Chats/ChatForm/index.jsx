import React from 'react';
import Modal from 'react-bootstrap/Modal';
import PropTypes from 'prop-types';
import { fetchUsersWithoutChat } from '../api/index';
import styles from './styles.module.scss';

class ChatForm extends React.Component {
  state = {
    users: [],
  }

  componentDidUpdate() {
    if (this.props.isShown && !this.state.users.length) this.fetchUsers();
    if (!this.props.isShown && this.state.users.length) this.setState({ users: [] });
  }

  renderUserRow = (user) => (
    <div className={styles.userItem} key={user.id} onClick={this.handleClick}>
      { user.full_name }
    </div>
  );

  render() {
    const { onClose, isShown } = this.props;
    return (
      <Modal
      show={isShown}
      onHide={onClose}
      size="md"
      aria-labelledby="contained-modal-title-vcenter"
      centered
    >
      <Modal.Header closeButton>
        <Modal.Title id="contained-modal-title-vcenter">
          Direct Messages
        </Modal.Title>
      </Modal.Header>
      <Modal.Body>
        { this.state.users.map((user) => this.renderUserRow(user)) }
      </Modal.Body>
    </Modal>
    );
  }

  handleClick = () => {
    console.log("Hello");
  }

  fetchUsers = () => {
    fetchUsersWithoutChat()
      .then((data) => {
        this.setState({ users: data });
      });
  }
}

ChatForm.propTypes = {
  onClose: PropTypes.func.isRequired,
  isShown: PropTypes.bool.isRequired,
};

export default ChatForm;
