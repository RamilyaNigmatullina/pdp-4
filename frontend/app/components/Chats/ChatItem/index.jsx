import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import moment from 'moment';
import styles from './styles.module.scss';

class ChatItem extends React.Component {
  renderAvatar = () => (
    <div className={styles.interlocutorAvatar}>
      <img
        className={classNames(styles.image, 'rounded-circle')}
        height="32"
        src={this.props.chat.interlocutor.avatar}
        width="32"
      />
    </div>
  );

  renderDate = () => (
    <div className={styles.date}>
      { moment(this.props.chat.last_message.created_at).calendar() }
    </div>
  );

  renderFullNameAndDate = () => {
    const { chat } = this.props;

    return (
      <div className={styles.details}>
        <div className={styles.interlocutorFullName}>
          { chat.interlocutor.full_name }
        </div>
        { chat.last_message && this.renderDate() }
      </div>
    );
  }

  renderUnreadMessagesBadge = () => (
    <div className={classNames(styles.badge, 'badge badge-pill badge-secondary')}>
      { this.props.chat.unread_messages_count }
    </div>
  );

  renderLastMessageAndUnreadMessagesBadge = () => {
    const { chat } = this.props;

    return (
      <div className={styles.details}>
        <div className={styles.lastMessage}>
          { chat.last_message && chat.last_message.text }
        </div>
        { !!chat.unread_messages_count && this.renderUnreadMessagesBadge() }
      </div>
    );
  }

  render() {
    const { chat, onClick, isCurrentChat } = this.props;
    const chatItemClasses = classNames(styles.chatItem, { [styles.active]: isCurrentChat });

    return (
      <div className={chatItemClasses} onClick={(e) => onClick(e, chat)}>
        <div className={styles.chatItemContainer}>
          { this.renderAvatar() }
          { this.renderFullNameAndDate() }
          { this.renderLastMessageAndUnreadMessagesBadge() }
        </div>
      </div>
    );
  }
}

ChatItem.propTypes = {
  isCurrentChat: PropTypes.bool.isRequired,
  chat: PropTypes.object.isRequired,
  onClick: PropTypes.func.isRequired,
};

export default ChatItem;
