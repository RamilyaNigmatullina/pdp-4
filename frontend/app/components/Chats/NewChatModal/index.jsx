import React from 'react';
import Modal from 'react-bootstrap/Modal';
import PropTypes from 'prop-types';
import { fetchUsersWithoutChat } from '../api/index';
import styles from './styles.module.scss';

class NewChatModal extends React.Component {
  state = {
    users: null,
  }

  componentDidUpdate() {
    if (this.props.isShown && !this.state.users) this.fetchUsers();
    if (!this.props.isShown && this.state.users) this.setState({ users: null });
  }

  renderUserRow = (user) => (
    <div className={styles.userItem} key={user.id} onClick={(e) => this.props.onChatCreate(e, user.id)}>
      { user.full_name }
    </div>
  );

  render() {
    const { onClose, isShown } = this.props;
    const { users } = this.state;

    if (!isShown) return null;

    return (
      <Modal aria-labelledby="contained-modal-title-vcenter" onHide={onClose} show={isShown} size="md" centered >
        <Modal.Header closeButton>
          <Modal.Title id="contained-modal-title-vcenter">
            Direct Messages
          </Modal.Title>
        </Modal.Header>
        <Modal.Body>
          { users && !!users.length && this.state.users.map((user) => this.renderUserRow(user)) }
          { users && !users.length && <div>No available users</div> }
        </Modal.Body>
      </Modal>
    );
  }

  fetchUsers = () => {
    fetchUsersWithoutChat().then((data) => this.setState({ users: data }));
  }
}

NewChatModal.propTypes = {
  onClose: PropTypes.func.isRequired,
  isShown: PropTypes.bool.isRequired,
  onChatCreate: PropTypes.func.isRequired,
};

export default NewChatModal;
