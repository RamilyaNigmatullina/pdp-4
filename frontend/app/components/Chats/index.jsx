import React from 'react';
import PropTypes from 'prop-types';
import consumer from 'helpers/consumer';
import ChatItem from './ChatItem';
import Chat from './Chat';
import NewChatModal from './NewChatModal';
import styles from './styles.module.scss';
import { readMessages, createChat } from './api/index';

class Chats extends React.Component {
  state = {
    chats: this.props.chats,
    currentChat: this.props.chats[0],
    isNewChatModalShown: false,
  };

  componentDidMount() {
    const { currentChat } = this.state;

    this.subscribeToChannel();

    if (currentChat && currentChat.unread_messages_count > 0) this.readChatMessages(currentChat);
  }

  subscribeToChannel = () => {
    consumer.subscriptions.create({
      channel: 'ChatNotificationsChannel',
    }, {
      received: this.handleMessageReceived,
    });
  }

  renderNewChatButton = () => (
    <button className="btn btn-outline-secondary btn-block" onClick={this.handleShowNewChatModal}>
      Start a new dialogue
    </button>
  );

  renderChatItem = (chat) => (
    <ChatItem
      chat={chat}
      isCurrentChat={this.isCurrentChat(chat)}
      key={chat.id}
      onClick={this.handleChatSelected} />
  );

  renderChat = () => (
    <Chat
      chat={this.state.currentChat}
      currentUser={this.props.currentUser}
      onMessageReceived={this.handleMessageReceived} />
  );

  renderChatsList = () => (
    <div className={styles.chatItems}>
      { this.renderNewChatButton() }
      { this.state.chats.map((chat) => this.renderChatItem(chat)) }
    </div>
  );

  render() {
    const { currentChat, isNewChatModalShown } = this.state;

    return (
      <div className={styles.chat}>
        <NewChatModal
          isShown={isNewChatModalShown}
          onChatCreate={this.handleCreateChat}
          onClose={this.handleCloseNewChatModal} />

        <div className={styles.chatsList}>
          { this.renderChatsList() }
        </div>
        <div className={styles.dialog}>
          { currentChat && this.renderChat() }
        </div>
      </div>
    );
  }

  handleMessageReceived = (chat) => {
    this.setState(() => {
      const originalChat = this.findOriginalChat(chat.id);
      const updatedChat = this.buildChat(originalChat, chat);

      if (this.isCurrentChat(updatedChat)) this.readChatMessages(updatedChat);

      return { chats: this.moveToTop(updatedChat) };
    });
  }

  handleChatSelected = (event, selectedChat) => {
    event.preventDefault();

    if (!this.isCurrentChat(selectedChat)) {
      if (!selectedChat.unread_messages_count) {
        this.setState({ currentChat: selectedChat });
      } else {
        this.readChatMessages(selectedChat);
      }
    }
  }

  handleShowNewChatModal = () => {
    this.setState({ isNewChatModalShown: true });
  }

  handleCloseNewChatModal = () => {
    this.setState({ isNewChatModalShown: false });
  }

  handleCreateChat = (event, userId) => {
    createChat(userId)
      .then((data) => {
        this.setState(({ chats }) => ({
          chats: [data, ...chats],
          currentChat: data,
          isNewChatModalShown: false,
        }));
      });
  }

  readChatMessages = (chat) => {
    readMessages(chat.id)
      .then(() => {
        const updatedChat = this.buildChat(chat, { unread_messages_count: 0 });

        this.setState({
          chats: this.replaceChat(updatedChat),
          currentChat: updatedChat,
        });
      });
  }

  findOriginalChat = (chatId) => this.state.chats.filter((chat) => chat.id === chatId)[0]

  buildChat = (originalChat, chat) => ({ ...originalChat, ...chat })

  moveToTop = (topChat) => [topChat, ...this.state.chats.filter((chat) => chat.id !== topChat.id)]

  replaceChat = (updatedChat) => this.state.chats.map((chat) => (chat.id === updatedChat.id ? updatedChat : chat))

  isCurrentChat = (chat) => {
    const { currentChat } = this.state;

    return !!currentChat && currentChat.id === chat.id;
  }
}

Chats.propTypes = {
  chats: PropTypes.array.isRequired,
  currentUser: PropTypes.object.isRequired,
};

export default Chats;
