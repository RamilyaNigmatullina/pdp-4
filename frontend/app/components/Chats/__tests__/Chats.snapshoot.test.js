import React from 'react';
import { shallow } from 'enzyme';
import Chats from '../index';

describe('Chats', () => {
  let props;

  const render = () => shallow(<Chats {...props} />);

  beforeEach(() => {
    props = {
      chats: [
        {
          id: 1,
          unread_messages_count: 0,
          interlocutor: {
            id: 2,
            full_name: 'Kevin Doe',
            avatar: '/kevin_doe_avatar.jpg',
          },
          last_message: {
            id: 1,
            text: 'Hello! How are you?',
            sender_id: 1,
            created_at: '2020-03-15T17:39:18.327Z',
          },
        },
        {
          id: 1,
          unread_messages_count: 2,
          interlocutor: {
            id: 3,
            full_name: 'Mark Boone',
            avatar: '/mark_boone_avatar.jpg',
          },
          last_message: {
            id: 1,
            text: 'Are you working?',
            sender_id: 1,
            created_at: '2020-03-15T15:00:10.327Z',
          },
        },
      ],
      currentUser: {
        id: 1,
        full_name: 'John Smith',
        avatar: '/john_smith_avatar.jpg',
      },
    };
  });

  it('renders chats', () => {
    const component = render();

    expect(component).toMatchSnapshot();
  });

  describe('without chats', () => {
    beforeEach(() => {
      props.chats = [];
    });

    it('renders chats', () => {
      const component = render();

      expect(component).toMatchSnapshot();
    });
  });
});
