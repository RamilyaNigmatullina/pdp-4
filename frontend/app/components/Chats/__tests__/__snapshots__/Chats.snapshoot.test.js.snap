// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chats renders chats 1`] = `
<div
  className="chat"
>
  <NewChatModal
    isShown={false}
    onChatCreate={[Function]}
    onClose={[Function]}
  />
  <div
    className="chatsList"
  >
    <div
      className="chatItems"
    >
      <button
        className="btn btn-outline-secondary btn-block"
        onClick={[Function]}
      >
        Start a new dialogue
      </button>
      <ChatItem
        chat={
          Object {
            "id": 1,
            "interlocutor": Object {
              "avatar": "/kevin_doe_avatar.jpg",
              "full_name": "Kevin Doe",
              "id": 2,
            },
            "last_message": Object {
              "created_at": "2020-03-15T17:39:18.327Z",
              "id": 1,
              "sender_id": 1,
              "text": "Hello! How are you?",
            },
            "unread_messages_count": 0,
          }
        }
        isCurrentChat={true}
        key="1"
        onClick={[Function]}
      />
      <ChatItem
        chat={
          Object {
            "id": 1,
            "interlocutor": Object {
              "avatar": "/mark_boone_avatar.jpg",
              "full_name": "Mark Boone",
              "id": 3,
            },
            "last_message": Object {
              "created_at": "2020-03-15T15:00:10.327Z",
              "id": 1,
              "sender_id": 1,
              "text": "Are you working?",
            },
            "unread_messages_count": 2,
          }
        }
        isCurrentChat={true}
        key="1"
        onClick={[Function]}
      />
    </div>
  </div>
  <div
    className="dialog"
  >
    <Chat
      chat={
        Object {
          "id": 1,
          "interlocutor": Object {
            "avatar": "/kevin_doe_avatar.jpg",
            "full_name": "Kevin Doe",
            "id": 2,
          },
          "last_message": Object {
            "created_at": "2020-03-15T17:39:18.327Z",
            "id": 1,
            "sender_id": 1,
            "text": "Hello! How are you?",
          },
          "unread_messages_count": 0,
        }
      }
      currentUser={
        Object {
          "avatar": "/john_smith_avatar.jpg",
          "full_name": "John Smith",
          "id": 1,
        }
      }
      onMessageReceived={[Function]}
    />
  </div>
</div>
`;

exports[`Chats without chats renders chats 1`] = `
<div
  className="chat"
>
  <NewChatModal
    isShown={false}
    onChatCreate={[Function]}
    onClose={[Function]}
  />
  <div
    className="chatsList"
  >
    <div
      className="chatItems"
    >
      <button
        className="btn btn-outline-secondary btn-block"
        onClick={[Function]}
      >
        Start a new dialogue
      </button>
    </div>
  </div>
  <div
    className="dialog"
  />
</div>
`;
